### ISSUE TITLE
Replication on Balabit: two feature bugs, a confound §5.2 misses, and a result that contradicts §7.3

### ISSUE THREAD

**Issue by BusqinTulga** [ck0i/SigmaDrift#1]
I rebuilt the missing `gen_cli`, reproduced the evaluation on Balabit, and attacked
the detector three times with changes *outside* the paper's adversarial search
space — not new parameter vectors, but different model structures. None evaded.
The third got EER from 0.036 to 0.209 and killed the linear detector outright, and
more importantly it broke the whack-a-mole constraint, which I think means §7.3
attributes the detector's robustness to the wrong cause.

Findings 1–3 are code-level and actionable regardless of what you make of the rest.

**TL;DR of the numbers** (5 held-out users, HGB, all generators emitting integer
coordinates event-driven):

| generator | AUROC | EER | TPR@1% | TPR@0.1% | AUROC (LR) |
|---|---|---|---|---|---|
| WindMouse | 0.9986 | 0.0091 | 0.991 | 0.635 | 0.981 |
| SigmaDrift upstream | 0.9920 | 0.0359 | 0.824 | 0.235 | 0.968 |
| my hand-built attempts | 0.9971 | 0.0260 | 0.929 | 0.823 | 0.953 |
| data-fitted shape model | **0.8634** | **0.2093** | 0.079 | 0.026 | **0.418** |

## Setup

- Human: Balabit `training_files`, 10 users, **5,216 aimed segments** via your own
  `preprocess_balabit.py` at defaults.
- Synthetic: condition-matched by resampling the empirical human joint distribution
  of (distance, direction, target_width).
- All 15 confounded features excluded. Leave-one-user-out, predictions pooled.
- XGBoost swapped for sklearn `HistGradientBoostingClassifier`
  (`max_depth=5, 200 iters, lr=0.05`) — no libomp available. LR is exactly your config.
- For the shape model the 10 users are split 5/5: the generator is fitted on one
  half, the detector only ever sees the other half. This matches your threat model
  (attacker has *some* public human data, defender has different data).

`gen_cli` reimplemented from the interface `src/adversarial.py` expects. Neither
repo ships it and §7 isn't reproducible without it — happy to PR.

---

## Finding 1: `smooth_curvature_change_rate` is ~99% a numerical singularity

Your strongest single feature — univariate AUROC 0.985–0.993 against every
generator I tried, top of Figure 7.

```python
speed_safe = np.where(speed > 1e-6, speed, 1e-6)
curvature  = cross / (speed_safe ** 3)          # 1e-6**3 == 1e-18
```

`np.gradient` uses central differences, so three consecutive identical positions
give `vx == vy == 0` up to floating-point residue. The residue survives into
`cross`, gets divided by 1e-18, and the feature returns ~1e9.

| | median κ' |
|---|---|
| human | **2.47** |
| upstream SigmaDrift | **1.89e10** |

Ten orders of magnitude. Human trajectories almost never contain duplicate
consecutive positions (2.9% of samples, mean longest run 1.0) because real mouse
input is **event-driven** — the input stack emits nothing when the cursor hasn't
moved. Synthetic resampled by `loader.resample_trajectory` is full of them (30–37%
of samples, runs of 5–10), because interpolation emits a sample whether or not the
position changed.

Dropping zero-displacement samples on the synthetic side:

| | median κ' | AUROC |
|---|---|---|
| as resampled | 1.89e10 | 0.985 |
| event-driven emit | **31** | **0.917** |

A real residual remains, so the feature measures *something*. As implemented it is
mostly a duplicate-sample detector. Suggest masking rather than clamping, and
dropping duplicate positions in the resampler.

I chased three wrong explanations before this one — stationary tails, pixel
quantization, insufficient smoothing. AUROC hid the mechanism completely; the raw
median exposed it instantly.

## Finding 2: `geo_curvature_integral` is substantially a pixel-quantization confound

Balabit records integer pixel coordinates; `generate()` returns doubles and the
detector consumed them as-is. Rounding synthetic to the integer grid — which is
what injected input events are anyway, you can't move the cursor to x=413.7:

| feature | float synthetic | rounded |
|---|---|---|
| geo_curvature_integral | 0.952 | **0.751** |
| sub_peak_count | 0.731 | 0.690 |
| smooth_curvature_change_rate | 0.993 | 0.988 |

Same category as the 15 excluded in §5.2.

Related, and worth a line in the generator: rounding each sample independently is
still not how a mouse works. The sensor accumulates sub-count displacement and
emits integer counts, so the cursor never reverses unless the true motion does.
Independent rounding of a smooth path dithers back and forth — sample-level
direction-reversal rate 0.096 vs 0.019 for humans. A five-line accumulator fixes it:

```cpp
double cx = 0, cy = 0;
for (auto& p : path) {
    double sx = std::trunc(p.x - cx), sy = std::trunc(p.y - cy);
    if (sx || sy) { cx += sx; cy += sy; emit(p.t, cx, cy); }
}
```

## Finding 3: the 17 "clean" features are not sampling-rate invariant

Holding the generator and the human set fixed and changing *only* the synthetic
emission density:

| target ISI | median samples | AUROC (HGB) | EER | AUROC (LR) |
|---|---|---|---|---|
| 10 ms | 97 | 0.9587 | 0.106 | 0.879 |
| 16 ms | 65 | 0.7775 | 0.303 | 0.531 |
| 30 ms | 38 | 0.9731 | 0.080 | 0.812 |
| 55 ms | 21 | 0.9666 | 0.096 | 0.772 |
| 90 ms | 14 | 0.8255 | 0.253 | 0.690 |

Stable across three seeds (isi=16: EER 0.294–0.303; isi=30: 0.074–0.080), so this
is a real, reproducible, non-monotonic dependence — a 4× swing in EER from
emission density alone.

§5.2 excludes the explicit timing and spectral families, but the smoothness and
curvature features are finite differences, so sampling rate re-enters through the
back door. Resampling synthetic to the human *median* ISI isn't sufficient when
the human ISI distribution is bimodal. Any reported margin is conditional on the
emission density chosen for the synthetic side, and that isn't stated.

## Finding 4: default Fitts constants aren't calibrated to anything

`config` ships `fitts_a=50, fitts_b=150`. Fitting Balabit with your own
preprocessing gives **a=446, b=192**. Over the observed ID range that nearly
coincides with the paper's a=200, b=260 — so §4.2's calibration looks right and
only the code default is off:

| | ID | human fit | paper (200/260) | code (50/150) |
|---|---|---|---|---|
| D=400 | 3.84 | 1185 ms | 1199 ms | **626 ms** |
| D=1000 | 5.10 | 1427 ms | 1526 ms | **815 ms** |

The shipped generator is ~2× too fast. Correcting it drops `fitts_mt` 0.833 → 0.578,
the largest single-feature improvement I found anywhere.

**Retracting a concern**: that `features.py` hardcoding `FITTS_A=50/FITTS_B=150`
while `--fit-fitts` defaults off could inflate detection. Tested both ways — EER
0.0026 vs 0.0025. The classifier also sees `fitts_mt` and `fitts_id`, so an affine
offset is redundant. Not a bug.

## Finding 5: Table 6's saturation hides real differences between generators

Table 6 reports XGBoost EER ≈ 0.0001 for SigmaDrift, SigmaDrift-retuned and
WindMouse alike. That reads as "all generators are equally detectable," and it's
what I initially took from it.

At a non-saturated operating point the ordering is clearly visible:

| | EER | TPR@0.1% |
|---|---|---|
| WindMouse | 0.0091 | 0.635 |
| SigmaDrift | 0.0359 | 0.235 |

SigmaDrift is ~4× harder to detect than WindMouse. The motor-control grounding
does buy something; Table 6 just can't show it. Worth reporting a metric that
isn't saturated, or the paper understates its own generator.

---

## Attack 1 — hierarchical population model

The generator is single-level: every call draws from the same fixed distributions,
and nearly every draw is `uniform(lo, hi)` (compactly supported, kurtosis 1.8, zero
tail mass by construction). The human pool is 10–29 self-consistent individuals.
No setting of Table 1 can synthesize a missing level.

At a fixed condition (D=400px, W=30px, n=20,000) — pure trial-to-trial variability:

| | upstream | patched |
|---|---|---|
| movement_time CV | **7.9%** | 30.9% |
| movement_time kurtosis | 3.09 | 4.79 |
| movement_time p1..p99 | [600, 868] ms | [339, 1434] ms |
| time_to_peak / MT CV | **5.5%** | 18.7% |

7.9% is `exp(N(0, 0.08))` on line 126, verbatim. The README lists "Fitts'
Compliance: Yes (~8%)" as a win, but ~8% residual scatter is *too* compliant.

**No detectability change.** EER 0.0025 → 0.0044. Widening variance around a wrong
mean actively hurt the Fitts family (0.864 → 0.931).

Two things I got wrong here, both arguing for your thesis:

- First attempt gave `ou_sigma`/`sdn_k` wide tails. Submovement counts exploded to
  53 peaks/trajectory — recognisably the round-5 "massive noise injection" strategy
  from Table 4, failing the same way.
- Correlating profile params against peak count (n=8,000): `primary_sigma` +0.507,
  `peak_time_ratio` +0.234, `ou_sigma` +0.055, `sdn_k` +0.047, `tremor_amp` +0.013.
  The *shape* parameter of the velocity primitive dominates; the noise gains are
  irrelevant. Timing and amplitude tolerate wide population spread, shape does not.

## Attack 2 — command-domain noise behind a limb model

`motor_synergy.h:231` adds OU drift, tremor and SDN **directly to position**. iid
position noise has a flat spectrum, so differentiating twice amplifies high
frequencies by ω² — unbounded jerk. A real hand has mass. This also misapplies the
cited source: Harris & Wolpert (1998) is about noise in the neural *command*, which
reaches the cursor only after muscle activation and limb inertia.

I moved all three terms into the command domain behind a damped second-order limb
model, so tremor emerges as narrowband ringing at resonance instead of being
injected as a fixed-phase pure tone.

**Worse.** The targeted feature went 0.985 → 0.997. Submovement features improved
sharply (`sub_peak_count` 0.687 → 0.519) and geometry degraded to compensate.
Whack-a-mole again. End-to-end EER 0.0260 — worse than upstream's 0.0359.

## Attack 3 — data-fitted shape model

Extract normalised progress/lateral profiles from real trajectories, PCA, fit a
conditional model on (log distance, direction harmonics), bootstrap residuals from
real trajectories at similar distance, reconstruct. Plus two refinements driven by
measured mechanism: transplanting real sub-pixel residual sequences index-wise
(interpolating them onto a different grid aliases them and makes things worse), and
accumulator emission.

Nearest-neighbour check in shape space: generated→training 0.325, held-out
human→training 0.420, training→training 0.419. Not replay, though generated
diversity sits ~23% below real.

| generator | AUROC | EER | TPR@1% | TPR@0.1% | AUROC (LR) |
|---|---|---|---|---|---|
| upstream | 0.9920 | 0.036 | 0.824 | 0.235 | 0.968 |
| shape model | 0.863–0.873 | **0.209–0.214** | 0.079 | 0.026 | **0.415–0.424** |

Three seeds. EER 5.8×; at a 1% false-ban rate the detector catches 8% of
trajectories instead of 82%; logistic regression is below chance.

**The per-feature picture is the part that matters:**

| feature | upstream | shape model |
|---|---|---|
| geo_curvature_integral | 0.828 | 0.792 |
| geo_path_efficiency | 0.844 | 0.689 |
| smooth_curvature_change_rate | 0.903 | 0.645 |
| sub_peak_count | 0.743 | 0.600 |
| fitts_residual | 0.866 | 0.564 |
| geo_max_deviation | 0.916 | 0.563 |
| fitts_mt | 0.814 | 0.546 |
| geo_angular_dev_at_peak | 0.915 | 0.529 |

**Every feature improved. None regressed.** Only one stays above 0.7, and that one
is partly the quantization confound from Finding 2. Every hand-built change I made,
including my own two attacks, produced the opposite pattern.

### Why this contradicts §7.3

§7.3 attributes the tradeoff to the generator's *architecture*: "the generator's
single parameter set creates a feed-forward trajectory where adjusting curvature to
better match geometry simultaneously distorts the jerk profile."

The shape model is also feed-forward and open-loop. It has no feedback, no reactive
corrections, nothing addressing §8.1. It nonetheless improved all four families at
once. The constraint is not feed-forward architecture — it is **model
misspecification**. When the model family doesn't contain the target distribution,
fitting one projection better necessarily worsens another. That is a property of
hand-specified parameter distributions, not of parametric generation.

The Optuna loop couldn't have found this because it searched 17 scalars over a
fixed structure. The whack-a-mole result bounds what parameter tuning can do, which
is narrower than the claim §7.3 draws from it.

### And §6.5's linear separability doesn't generalize

Against the shape model, logistic regression gets AUROC 0.418. §6.5 infers from
LR ≈ XGBoost that "the human–synthetic separation is approximately linearly
separable in this feature space." That holds for SigmaDrift, not for the feature
space.

### Honest attribution, since it undercuts the result

| component | EER | gain | nature |
|---|---|---|---|
| upstream | 0.036 | — | |
| + learned shape model | 0.080 | 2.2× | genuine synthesis |
| + residual re-injection | 0.189 | **2.4×** | **transplanted real micro-structure** |
| + accumulator emission | 0.209 | 1.1× | emission-layer fix |

The largest single lever is transplanting real recorded sub-pixel residual
sequences — the component closest to replay. The accumulator applied to upstream
SigmaDrift on its own is worth only 1.2× (0.036 → 0.043), so the emission fix is
not carrying the result, but the residual transplant is doing more work than the
learned model is.

The shape model is also data-bound (needs thousands of matched trajectories from
the target context, no physiology to fall back on), context-bound (fitted to
Balabit desktop browsing at ~30 ms polling), ~23% under-dispersed relative to real
humans, and untested against a defender who normalises sampling density — which
Finding 3 says would matter. It is a research artifact, not a better generator.

---

## Where this leaves the paper

- **Central claim survives, with less margin than reported.** Three structural
  attacks, none evaded. Best result is EER 0.209, AUROC 0.87; over a 5-movement
  window at 1% FPR that is still ~39% detection. But "EER ≤ 0.001 against the
  strongest parametric adversary we could construct" does not survive contact with
  a generator fitted to data, and part of the reported margin is Findings 1–3.
- **§7.3's mechanism is wrong.** The tradeoff is misspecification, not feed-forward
  architecture. A feed-forward model fitted to data doesn't exhibit it.
- **§6.5's linear separability is generator-specific**, not a property of the
  feature space.
- **§5.2 is the paper's best contribution and is incomplete.** Findings 1–3 are
  three more confound classes in the same spirit, and one of them is the top
  feature by SHAP importance.

## Caveats

Balabit only (10 users; 5/5 split for Attack 3), no BOUN, so absolute numbers
aren't comparable to Table 2 and the human pool is less diverse than yours.
HistGradientBoosting substitutes for XGBoost. Target width is the fabricated 30px
constant your preprocessor assumes, so all Fitts numbers inherit that. My `gen_cli`
and condition matching are reimplementations and may differ from yours. Attack 3's
train and test users share a collection protocol; cross-dataset generalisation is
untested.

Nothing here touches §8.1 — every generator tested is open-loop with pre-computed
corrections. Attack 3 suggests that may matter less than §7.3 implies, since a
purely feed-forward model already removes the tradeoff.

Patch, scripts and the shape model available if useful. Happy to PR `gen_cli`, the
singularity guard, the resampler dedup and the accumulator emission independently
of what you make of the rest.

**Comment by ck0i** [comment]
Thanks for taking the time to investigate this. The curvature and sampling concerns look worth checking, please share the scripts and patches so I can reproduce your results.

I do disagree with how you’ve framed the contradictions. §7.3 explicitly calls the architectural explanation a hypothesis and acknowledges that a more expressive parametric model could avoid those tradeoffs. §6.5 also explicitly limits the linear-separability observation to the datasets and generators tested. Your results could help explore those limits, but those qualifications are already in the paper.

One detail I’d like clarified: the published evaluation code treats human as the positive class. Did your harness reverse both labels and scores before interpreting TPR@1% FPR as bot detection at a 1% false-ban rate?

Given the different classifier, smaller population, and reconstructed harness, I’d want a matched comparison before drawing conclusions about the reported detection margin. I’m happy to review reproducible fixes; I just think the broader conclusions need more support than the tables alone provide.

**Comment by ck0i** [comment]
No response given. Closing this issue.
